feat(compass-query-bar): add query ai feedback options COMPASS-7051

Anemy opened the pull request:
COMPASS-7051

Adds two buttons when there was a successful ai query generation. Submitting the feedback emits a telemetry event with the message.



https://github.com/mongodb-js/compass/assets/1791149/75bf03c1-f7b5-4182-8aad-c24332eb90ef

<img width="290" alt="Screenshot 2023-08-09 at 2 15 45 PM" src="https://github.com/mongodb-js/compass/assets/1791149/97bd196b-22f5-4fce-8f2e-4a26dcf955e4">

| light | dark |
| - | - |
| <img width="432" alt="Screenshot 2023-08-09 at 2 15 54 PM" src="https://github.com/mongodb-js/compass/assets/1791149/b1c2e8a3-7afd-4fcd-b112-e7b587b82b70"> | <img width="444" alt="Screenshot 2023-08-09 at 2 16 05 PM" src="https://github.com/mongodb-js/compass/assets/1791149/039bf525-5952-4450-bd5d-03fa0187c2db"> |
| <img width="403" alt="Screenshot 2023-08-09 at 2 15 50 PM" src="https://github.com/mongodb-js/compass/assets/1791149/778cadd7-ee17-40e5-8a7c-3ac9d027e0d8"> | <img width="393" alt="Screenshot 2023-08-09 at 2 16 10 PM" src="https://github.com/mongodb-js/compass/assets/1791149/45eefc9f-637f-4348-8e66-16d85b3685fb"> |

gribnoysup commented on packages/compass-components/src/components/leafygreen.tsx:
```diff
@@ -102,6 +103,10 @@ export {
   Icon,
   IconButton,
   Language,
+  // We export the GuideCue as LGGuideCue as we have our own wrapped
+  // GuideCue implementation in Compass. LGGuideCue is only to be
+  // used in situations where Compass' GuideCue is not intended for use.
+  LGGuideCue,
```
This will be eventually used for aggregation right? Maybe worth putting this component to compass-components as some sort of FeedbackTooltip instead of exposing lg guide cue like that?

gribnoysup replied:
No strong preference on this! Just thinking how we can keep GuideCue internal 🤔

Anemy replied:
Good thinking, updated. It does still feel like GuideCue is maybe not the right component for this from LeafyGreen, I think it's better than custom popovers for now, we can revisit down the line I reckon.
I went with FeedbackPopover instead of FeedbackTooltip if that's alright with you, I can switch it back if you prefer.

gribnoysup replied:
No, no preference, both make sense, I guess we could use an InteractivePopover, but it doesn't have a tooltip pointer thingy, right?

Anemy replied:
No tooltip pointer thing yup. Let's see if we end up doing more of this sort of thing in which case we can add something to the interactive popover for that and use it here.

gribnoysup commented on packages/compass-components/src/components/feedback-popover.tsx:
```diff
@@ -0,0 +1,52 @@
+  LGGuideCueProps,
+  'currentStep' | 'numberOfSteps' | 'children' | 'title'
+> & {
+  onFeedback: (text: string) => void;
```
Tiny naming nit: maybe `onFeedbackSubmit` or just `onSubmit`? I think the part about this being feedback can be inferred from the component name

gribnoysup commented on packages/compass-query-bar/src/components/generative-ai/query-feedback.tsx:
```diff
@@ -0,0 +1,118 @@
+});
+
+const feedbackButtonLightStyles = css({
+  // We fill the icon colors here as there is a bug with the
+  // LeafyGreen ThumbsUp and ThumbsDown icons.
```
Is there a ticket we can link to here?

Anemy replied:
Added one! Didn't have at the time of writing, thanks for catching.

gribnoysup commented on packages/compass-query-bar/src/stores/ai-query-reducer.ts:
```diff
@@ -22,6 +22,7 @@ export type AIQueryState = {
   aiPromptText: string;
   status: AIQueryStatus;
   aiQueryFetchId: number; // Maps to the AbortController of the current fetch (or -1).
+  didSubmitFeedback: boolean;
```
Don't have the strongest preference, but as this basically relies on query succeeding (state we already have in store!) and telemetry (`track` calls) for actual feedback submission, this whole thing including its state can live outside of the redux store? As I see it, you can track submission state inside the `QueryFeedback` component and (un)mounting it when ai query state changes should be enough to correctly reset it, so something like this:

```tsx
function AITextInput() {
  // ...

  {didSucceed && <QueryFeedback />}
}

...

function QueryFeedback() {
  const [didSubmit, setDidSubmit] = useState(false);
  
  const onSubmit = () => {
    track(...)
    setDidSubmit(true);
  }

  if (didSubmit) {
    return null; // ... or a fancy animation if we want to now that we control the rendering from here :)
  }
  
  // ...
}
```

I think this would have two big benefits: less state in redux store (always better if can be done!) and probably would be easier to share it with agg builder in the future

Anemy replied:
Done! Also added an animation for when the feedback is submitted:

https://github.com/mongodb-js/compass/assets/1791149/cb659813-2e2b-4892-9f45-fcda8a4ca601

gribnoysup reviewed (approved):
Generally looks good, but I think we can isolate this form from redux state completely

gribnoysup commented:
Oh, actually, a couple of things I spotted when double-checking locally: 

- When switching top level tabs, the feedback form stays on the screen. Probably another case where we don't unmount something causing this. We can either start unmouting top level tabs (but who knows what code relies on this, hopefully none, but risky) or you can use the same intersection observer behavior we use in other places to handle similar cases

![image](https://github.com/mongodb-js/compass/assets/5036933/c41885ad-06af-4b02-9e47-accacb0abbf5)

- There doesn't seem to be a way to dismiss the form without submitting, but you can submit an empty response (not sure if that's the expected behavior). On top of that pressing `Esc` also submits a form and this button should probably be more of an indicator that the form got dismissed instead of submitted?

Anemy commented:
@gribnoysup Thanks for catching the tab switching behavior. Updated it to dismiss when clicked outside. Also updated the component to close with `Escape` and submit with `Enter`.

gribnoysup reviewed (approved):
Nice handling of the keyboard events. Two small things I noticed, otherwise looks good
